src/TextForEvent.tsx

```diff
@@ -33,14 +33,83 @@ import { ROOM_SECURITY_TAB } from "./components/views/dialogs/RoomSettingsDialog
 
 import { logger } from "matrix-js-sdk/src/logger";
 import { removeDirectionOverrideChars } from 'matrix-js-sdk/src/utils';
+import CallHandler from './CallHandler';
+import { RoomMember } from 'matrix-js-sdk/src/models/room-member';
+
+/*
+ * Attempt to retrieve the displayname of a user from an event. If the user is a virtual user,
+ * return the displayname of the native user instead. In either case, if a displayname cannot
+ * be found, simply return the MXID of the user instead.
+ *
+ * @param event The event to extract the user from.
+ */
+function getSenderNameFromPotentiallyVirtualUser(event: MatrixEvent) {
+    // Retrieve the room ID from the event
+    const potentiallyVirtualRoom = MatrixClientPeg.get().getRoom(event.getRoomId());
+
+    // Check if the event is from a virtual room
+    if (
+        !CallHandler.sharedInstance().getSupportsVirtualRooms() ||
+        !window.mxVoipUserMapper.isVirtualRoom(potentiallyVirtualRoom)
+    ) {
+        // If not, simply extract the sender information from the incoming event
+        return event.sender ? event.sender.name : _t('Someone');
+    }
+
+    // Otherwise, assume the caller is a virtual user and attempt to look up the corresponding
+    // native user
+
+    // We can avoid a potentially expensive virtual -> native user lookup by checking if the native room
+    // only has two people in it. If so, then the other user in that room is likely the matching native user.
+    const nativeRoomId = window.mxVoipUserMapper.nativeRoomForVirtualRoom(potentiallyVirtualRoom.roomId);
+    const nativeRoom = MatrixClientPeg.get().getRoom(nativeRoomId);
+    const nativeRoomjoinedMembers = nativeRoom.getJoinedMembers();
+
+    const getSenderNameForUserId = (userId) => {
+        return MatrixClientPeg.get().getProfileInfo(userId).then((resp) => {
```

Comment:
it doesn't matter in the long run, but we try to use `async/await` in newer code to avoid confusing promise semantics - up to the developer on whether that is clearer or not though